compiler/packages/babel-plugin-react-compiler/src/HIR/HIR.ts

```diff
@@ -1429,6 +1429,8 @@ export type ReactiveScope = {
   merged: Set<ScopeId>;
 
   loc: SourceLocation;
+
+  source: boolean;
```

Comment:
We probably should check `source` in all passes that act on scopes.

```js
// flattenScopesWithHooksOrUse
function useFoo() {
  // Assume that useARandomFunction is not a hook
  const res = useMemo(() => useARandomFunction(), []);
  return res;
}
```

```js
// flattenReactiveLoops
function useFoo() {
  const res = []
  for (let i = 0; i < 5; i++) {
    res.push(useMemo(() => foo(), []));
  }
  return res;
}
```

(etc)